File: backend/apps/compras/presentation/permissions.py
"""Django permission codenames for compras recebimento models."""

from rest_framework.request import Request

APP = "compras_infrastructure"

VIEW_FORNECEDOR = f"{APP}.view_fornecedor"
ADD_FORNECEDOR = f"{APP}.add_fornecedor"
CHANGE_FORNECEDOR = f"{APP}.change_fornecedor"
DELETE_FORNECEDOR = f"{APP}.delete_fornecedor"

VIEW_FORNEC_CONTATO = f"{APP}.view_forneccontato"
ADD_FORNEC_CONTATO = f"{APP}.add_forneccontato"
CHANGE_FORNEC_CONTATO = f"{APP}.change_forneccontato"
DELETE_FORNEC_CONTATO = f"{APP}.delete_forneccontato"

VIEW_MSG_ERRO = f"{APP}.view_msgerro"

VIEW_PAIS = f"{APP}.view_pais"


def _has_pk(request: Request, field: str) -> bool:
    value = request.data.get(field)
    return value is not None and value != ""


def fornecedor_list_or_grava_perms(request: Request) -> list[str]:
    if request.method == "GET":
        return [VIEW_FORNECEDOR]
    if _has_pk(request, "cod_fornec"):
        return [CHANGE_FORNECEDOR]
    return [ADD_FORNECEDOR]


def fornec_contato_list_or_grava_perms(request: Request) -> list[str]:
    if request.method == "GET":
        return [VIEW_FORNEC_CONTATO]
    if _has_pk(request, "cod_contato"):
        return [CHANGE_FORNEC_CONTATO]
    return [ADD_FORNEC_CONTATO]


def fornec_contato_get_or_delete_perms(request: Request) -> list[str]:
    if request.method == "DELETE":
        return [DELETE_FORNEC_CONTATO]
    return [VIEW_FORNEC_CONTATO]

File: frontend/src/modules/purchasing/permissions.ts
/** Django permission strings for Compras models (`app_label.codename`). */
export const PURCHASING_PERMS = {
  viewFornecedor: "purchasing_infrastructure.view_fornecedor",
  viewDashboard: "purchasing_infrastructure.view_dashboard",
  addFornecedor: "purchasing_infrastructure.add_fornecedor",
  changeFornecedor: "purchasing_infrastructure.change_fornecedor",
  deleteFornecedor: "purchasing_infrastructure.delete_fornecedor",
  viewFornecContato: "purchasing_infrastructure.view_forneccontato",
  addFornecContato: "purchasing_infrastructure.add_forneccontato",
  changeFornecContato: "purchasing_infrastructure.change_forneccontato",
  deleteFornecContato: "purchasing_infrastructure.delete_forneccontato",
} as const;

export type PurchasingPermission = (typeof PURCHASING_PERMS)[keyof typeof PURCHASING_PERMS];
